### PR TITLE
Enable multiple passes over the dataset.

### PR DESCRIPTION
**Pull request by GeorgiosSmyrnis**
This PR adds an extra flag `--multiple-data-passes` which allows for multiple passes over the dataset.
The code is set up so that the total number of tokens that will be seen across all passes of the dataset will still be `epochs * train_num_samples`, only in this case instead of erroring when data runs out, we move on to the next pass of the dataset (shuffling the shards as well, so that we see them in different order in each pass).

This is a WIP, since a few more test cases need to be added.

### CODE REVIEW

**Review by Vaishaal — APPROVED** [review]

**Comment by xszheng2020** [comment]
Hi, @GeorgiosSmyrnis 

I am training a dclm 400m-4x model using a limited amount of tokens like 1B tokens allowing multiple passes.
Then I observe the training loss curve as follows. 
The shape looks strange since the training loss drops periodically. 
Could it be because we did not shuffle the shards correctly?

Any idea with this?

Best,
Xiaosen

image
